### PR TITLE
mixminus: add conference feature

### PR DESCRIPTION
**Pull request by sreimers**

### CODE REVIEW

**Comment by cspiel1** [comment]
- RFCs? https://tools.ietf.org/html/rfc4353
- Mixer: mixausrc mixes encode/decode stream with an audio file. Maybe it can help to find some ideas for a conference mixer.

File: src/audio.c [hunk 1]
```diff
@@ -2562,6 +2563,38 @@ void audio_set_hold(struct audio *au, bool hold)
...
+/**
+ * Is audio on conference?
+ *
+ * @param au    Audio object
+ *
+ * @return true if on conference, false if not
+ */
+bool audio_is_conference(const struct audio *au)
+{
+	return au->conference ? true : false;
```
Comment (alfredh) [inline src/audio.c:2594]: suggested code:

```c
return au ? au->conference : false;
```

File: modules/mixminus/mixminus.c [hunk 2]
```diff
@@ -0,0 +1,401 @@
...
+				return EINVAL;
+			}
+		}
+		else {
+			aubuf_read_samp(mix->ab, sampv_mix, af->sampc);
+		}
+
+		for (i = 0; i < af->sampc; i++) {
+			sampv[i] = sampv[i] + sampv_mix[i];
+		}
```
Comment (alfredh) [inline modules/mixminus/mixminus.c:277]: did you consider saturation of added samples here ?

the range is 16-bit signed, if two added samples have e.g. value 32000 and 32000,
the sum will be 64000 and outside the range.

it is possible to add the samples using a 32-bit integer and then soft clip at +/- 32768.

File: src/call.c [hunk 3]
```diff
@@ -1243,6 +1243,25 @@ int call_hold(struct call *call, bool hold)
...
+ */
+int call_conference(struct call *call, bool conference)
+{
+	struct audio *au = call_audio(call);
+
+	if (!au)
+		return EINVAL;
+
+	return audio_set_conference(au, conference);
+}
```
Comment (alfredh) [inline src/call.c:1262]: also consider using the `call_audio` function to fetch the audio object from the call object.
then there is no need for an explicit wrapper function.

**Comment by alfredh** [comment]
did you consider using the `aumix` module in rem:

```c
int aumix_alloc(struct aumix **mixp, uint32_t srate,
		uint8_t ch, uint32_t ptime);
```

Btw, when you mention "High Latency" how many milliseconds/seconds are we talking about ?

File: modules/mixminus/mixminus.c (hunk 2 above)
Comment (sreimers) [reply modules/mixminus/mixminus.c:277]: Yes I`m aware of this, still in proof of concept stage, will fix this.

**Comment by sreimers** [comment]
> did you consider using the `aumix` module in rem:
> 
> ```c
> int aumix_alloc(struct aumix **mixp, uint32_t srate,
> 		uint8_t ch, uint32_t ptime);
> ```

Had a look at it, but I think its not flexible enough and adds a extra thread per call to keep in sync.


> Btw, when you mention "High Latency" how many milliseconds/seconds are we talking about ?

I have fixed latency and syncing problems now.

**Comment by alfredh** [comment]
should we update the patch to use the original aufilt api ?

looks like there is an error from the build:

```
modules/mixminus/mixminus.c:456:18: error: nested extern declaration of ‘baresip_au_pre_filtl’ [-Werror=nested-externs]
516
```

**Pull request merged**